From 2bc7f026d1bfda71826225572721768cd161b1ff Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 25 Jan 2021 13:25:25 +0200
Subject: [PATCH] [SW-30229] ppboot: goya2: set timeout and enable drain

Currently while running workloads, it is observed that some of
blocks register becomes inaccessible from ARC due to error on block
which causes ARC to go in bad state as the transaction to inaccessible
block doesn't get completed. Upon hard reset of device after failure
and even after performing halt and reset of ARC as part of ppboot,
the ARC doesn't come in good state and doesn't respond to halt request.

This patch add changes to enable drain control and set timeout of 1000
cycles which invalidates incomplete transaction after timeout. Upon
performing above configuration, ARC is able to respond to halt request
in error scenario.

Change-Id: Iaac94c2e7d49776526690cae30d3d8bf69c3ec95
---

diff --git a/goya2/common.c b/goya2/common.c
index e0c376d..b6533ba 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -407,8 +407,22 @@
 
 }
 
+static void axi_enable_drain(void)
+{
+	u32 val;
+
+	/* Set timeout and enable drain */
+	regmap_writel(PSOC_AXI_DRAIN_TIMEOUT,
+			mmPSOC_GLOBAL_CONF_AXI_DRAIN_TIMEOUT);
+	val = regmap_readl(mmPSOC_GLOBAL_CONF_AXI_DRAIN_CTRL);
+	val |= (0x1 << PSOC_GLOBAL_CONF_AXI_DRAIN_CTRL_EN_SHIFT);
+	regmap_writel(val, mmPSOC_GLOBAL_CONF_AXI_DRAIN_CTRL);
+}
+
 void hl_plat_init(void)
 {
+	axi_enable_drain();
+
 	init_arc();
 }
 
diff --git a/goya2/common.h b/goya2/common.h
index 4fa1b72..247c9bb 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -99,4 +99,5 @@
 #define PCIE_PHY_PROTOCOL0_EXT_RX_ADAPT_MODE_G3_CONF	0x5555
 #define PCIE_PHY_PROTOCOL0_EXT_RX_ADAPT_SEL_G4_CONF	0xFF
 
+#define PSOC_AXI_DRAIN_TIMEOUT				0x3E8 /* 1000 cycles */
 #endif /* COMMON_GOYA2_H_ */
